DeviceBridge/Services/SubscriptionCallbackFactory.cs

```diff
@@ -0,0 +1,203 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+
+using System;
+using System.IO;
+using System.Net.Http;
+using System.Text;
+using System.Text.Json;
+using System.Threading.Tasks;
+using DeviceBridge.Models;
+using Microsoft.Azure.Devices.Client;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using NLog;
+
+namespace DeviceBridge.Services
+{
+    /// <summary>
+    /// This module contains the logic to build custom device client callbacks for subscriptions.
+    /// The callbacks convert C2D/connection events into HTTP notifications.
+    /// </summary>
+    public class SubscriptionCallbackFactory : ISubscriptionCallbackFactory
+    {
+        private readonly Logger _logger;
+        private readonly IHttpClientFactory _httpClientFactory;
+
+        public SubscriptionCallbackFactory(Logger logger, IHttpClientFactory httpClientFactory)
+        {
+            _logger = logger;
+            _httpClientFactory = httpClientFactory;
+        }
+
+        public DesiredPropertyUpdateCallback GetDesiredPropertyUpdateCallback(string deviceId, DeviceSubscription desiredPropertySubscription)
+        {
+            return async (desiredPopertyUpdate, _) =>
+            {
+                _logger.Info("Got desired property update for device {deviceId}. Callback URL: {callbackUrl}. Payload: {desiredPopertyUpdate}", deviceId, desiredPropertySubscription.CallbackUrl, desiredPopertyUpdate.ToJson());
+
+                try
+                {
+                    var body = new DesiredPropertyUpdateEventBody()
+                    {
+                        DeviceId = deviceId,
+                        DeviceReceivedAt = DateTime.UtcNow,
+                        DesiredProperties = new JRaw(desiredPopertyUpdate.ToJson()),
+                    };
+
+                    var payload = new StringContent(JsonConvert.SerializeObject(body), Encoding.UTF8, "application/json");
+                    using var httpResponse = await _httpClientFactory.CreateClient("RetryClient").PostAsync(desiredPropertySubscription.CallbackUrl, payload);
+                    httpResponse.EnsureSuccessStatusCode();
+                    _logger.Info("Successfully executed desired property update callback for device {deviceId}. Callback status code {statusCode}", deviceId, httpResponse.StatusCode);
+                }
+                catch (Exception e)
+                {
+                    _logger.Error(e, "Failed to execute desired property update callback for device {deviceId}", deviceId);
```

Comment:
Looks like we ate callback exceptions, what will API caller get? do they get a failure notification?

Reply:
This only happens after we retried the notification multiple times and it failed. There's nothing we can do in this case, as the user webhook endpoint is unhealthy, so the only thing we can do is log the failure.
(btw, this entire file was just moved from `SubscriptionService` the code is all the same it was before)